syntax = "proto3";

package raft;

// Raft RPC service definition
service RaftService {
  // RequestVote RPC - called by candidates to gather votes
  rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
  
  // AppendEntries RPC - called by leader to replicate log entries and as heartbeat
  rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
  
  // InstallSnapshot RPC - called by leader to send snapshot to follower
  rpc InstallSnapshot(InstallSnapshotRequest) returns (InstallSnapshotResponse);
}

// RequestVote RPC arguments
message RequestVoteRequest {
  int32 term = 1;              // candidate's term
  string candidate_id = 2;      // candidate requesting vote
  int32 last_log_index = 3;     // index of candidate's last log entry
  int32 last_log_term = 4;      // term of candidate's last log entry
}

// RequestVote RPC response
message RequestVoteResponse {
  int32 term = 1;               // currentTerm, for candidate to update itself
  bool vote_granted = 2;        // true means candidate received vote
}

// AppendEntries RPC arguments
message AppendEntriesRequest {
  int32 term = 1;               // leader's term
  string leader_id = 2;         // so follower can redirect clients
  int32 prev_log_index = 3;     // index of log entry immediately preceding new ones
  int32 prev_log_term = 4;      // term of prevLogIndex entry
  repeated LogEntry entries = 5; // log entries to store (empty for heartbeat)
  int32 leader_commit = 6;      // leader's commitIndex
}

// AppendEntries RPC response
message AppendEntriesResponse {
  int32 term = 1;               // currentTerm, for leader to update itself
  bool success = 2;             // true if follower contained entry matching prevLogIndex and prevLogTerm
  int32 conflict_index = 3;     // index of first conflicting entry (optimization)
  int32 conflict_term = 4;      // term of conflicting entry (optimization)
}

// Log entry structure
message LogEntry {
  int32 term = 1;               // term when entry was received by leader
  int32 index = 2;              // position in the log
  string command = 3;           // command for state machine (JSON-encoded)
  string command_type = 4;      // Type of command: RATE_LIMIT_CHECK, RATE_LIMIT_RESET, etc.
}

// InstallSnapshot RPC arguments
message InstallSnapshotRequest {
  int32 term = 1;               // leader's term
  string leader_id = 2;         // so follower can redirect clients
  int32 last_included_index = 3; // snapshot replaces all entries up through this
  int32 last_included_term = 4;  // term of last_included_index
  bytes data = 5;               // raw bytes of the snapshot
}

// InstallSnapshot RPC response
message InstallSnapshotResponse {
  int32 term = 1;               // currentTerm, for leader to update itself
}
